<!DOCTYPE html>
<html lang="zh-cn">
	<head>
		<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
        <title></title>
		<style>
			
		</style>
		<script>
			
			;(function (){
				// 定义三种状态
				let status = {
					pending:0,
					fulfilled:1,
					rejected: 2
				}
				class CustomePromise {
					constructor(executor){
						this._status = status.pending;
						this._value = null;

						this.fulfillQueues = [];
						this.rejectQueues = [];
						
						this._handler(executor)
						
					}

					_handler(executor){
						let done = false;
						try{
							executor((value) => {
								if(done) return;
								done = true;
								this._resolve(value)
							}, (error) => {
								if(done) return;
								done = true;
								this._reject(error)	
							})
						}catch(e){
							this._reject(e);
						}
					}

					_resolve(value){
						// 调用then传递的成功执行的方法....
						setTimeout( () => {

							let then = this._getThen(value);
							if(then){
								return this._handler(then.bind(value))
							}
							this._status = status.fulfilled;
							this._value = value;
							this.fulfillQueues.forEach(item=>item(this._value))	
						});
						
					}

					_reject(error){
						// 调用then传递的失败执行的方法....
						setTimeout( () => {
							this._status = status.rejected;
							this._value = error;
							this.rejectQueues.forEach(item=>{
								item(this._value)
							})	
						});
					}

					_getThen(value){
						let type = typeof value;
						if(value && (type === 'object' || type==='function')){
							let then;
							if(then = value.then){
								return then;
							}
						}

						return null;
					}

					then(fulfillFunc,rejectFunc){
						return new CustomePromise( (resolve,reject) => {
							// 收集成功和失败回调

							let fulfillFuncNext = (value) => {
								
								try{
									if(typeof fulfillFunc !== 'function'){
										resolve(value);
										return;
									}
									let returnValue = this._getThen(fulfillFunc(value));
									if(returnValue){
										this._handler(returnValue.bind(value))
										return;
									}
									resolve(returnValue);
								}catch(e){
									reject(e)
								}
							}
							let rejectFuncNext = () => {
								
							}

							if(this._status === 0){
								this.fulfillQueues.push(fulfillFuncNext);
								this.rejectQueues.push(rejectFuncNext);
							}else if(this._status === 1){
								fulfillFunc(this._value);
							}else if(this._status === 2){
								rejectFunc(this._value);
							}
						});
					}
				}	

				window.CustomePromise = CustomePromise;
			})();

			/*
				resolve的值有三种：
					1. 普通的值
					2. Promise对象
					3. thenable对象，{then(){}}
			*/

			let p0 = new CustomePromise((resolve,reject) => {
				setTimeout(function (){
					resolve('p0')	
				},3000)
			});

			p0.then((data) => {
				console.log('p0.thenp0p0p0p0',data);
			},function (data){
				console.log('p0.then2',data);	
			}).then((data) => {
				console.log('p0.thenp0p0p0p0',data);
			},function (data){
				console.log('p0.then2',data);	
			})

			p0.then((data) => {
				console.log('p0.then2222',data);
			},function (data){
				console.log('p0.then22222',data);	
			})

			let p = new CustomePromise((resolve,reject) => {
				setTimeout(function (){
					resolve(p0);
				},1000)
			})

			p
				.then(function(data){
					console.log(this,1111111111);
					console.log('第一个then的成功函数,数据为',data);
					return new CustomePromise((resolve,reject) => {
						setTimeout(function (){
							resolve('延时了一下')	
						},1000)
					});
				},function (error){
					console.log('第一个then的失败函数,数据为',error);	
				})
				.then(function (data){
					console.log('第2个then的成功函数,数据为',data);	
				},function (error){
					console.log('第2个then的失败函数,数据为',error);		
				});

			p.then((data) => {
				console.log(123,data);
			})

			console.log('我先执行');
		</script>
	</head>

	<body>
		
	</body>
</html>